reco_utils/dataset/pandas_df_utils.py

```diff
@@ -166,3 +166,93 @@ def _convert(field, feature, field_index, field_feature_index_dict):
     return df_new
 
 
+def negative_feedback_sampler(
+    df, 
+    col_user=DEFAULT_USER_COL,
+    col_item=DEFAULT_ITEM_COL,
+    number_neg_per_pos=1,
+    seed=42
+):
+    """Utility function to sample negative feedback from user-item interaction dataset.
+
+    This negative sampling function will take the user-item interaction data to create 
+    binarized feedback, i.e., 1 and 0 indicate positive and negative feedback, 
+    respectively. 
+
+    Negative sampling is used in the literature frequently to generate negative samples 
+    from a user-item interaction data.
+    See for example the neural collaborative filtering paper 
+    https://www.comp.nus.edu.sg/~xiangnan/papers/ncf.pdf
+    
+    Examples:
+        >>> import pandas as pd
+        >>> df = pd.DataFrame({
+            'userID': [1, 2, 3],
+            'itemID': [1, 2, 3],
+            'rating': [5, 5, 5]
+        })
+        >>> df_neg_sampled = negative_sampler_from_user_item_tuple(
+            df, col_user='userID', col_item='itemID', number_neg_per_pos=1
+        )
+        >>> df_neg_sampled
+        userID  itemID  feedback
+        1   1   1
+        1   2   0
+        2   2   1
+        2   1   0
+        3   3   1
+        3   1   0
+
+    Args:
+        df (pandas.DataFrame): input data that contains user-item tuples.
+        col_user (str): user id column name.
+        col_item (str): item id column name.
+        number_neg_per_pos (int): number of negative feedback sampled for each positive feedback on the user-item interactive data. 
+        seed (int): seed for the random state of the sampling function.
+
+    Returns:
+        pandas.DataFrame: data with negative feedback 
+    """
+    columns = df.columns
+    if col_user not in columns:
+        raise ValueError("Column {} is not in the dataframe".format(col_user))
+
+    if col_item not in columns:
+        raise ValueError("Column {} is not in the dataframe".format(col_item))
+
+    # Get all of the users and items.
+    users, items = df[col_user].unique(), df[col_item].unique()
+
+    # Create a dataframe for all user-item pairs
+    user_item_tuples = [(row[col_user], row[col_item]) for _, row in df.iterrows()]
+
+    # Generate feedback for both positive and negative
+    user_item_feedback = [
+        (user, item, 1 if (user, item) in user_item_tuples else 0) for user in users for item in items
```

Comment:
Why do you need the Cartesian product of users and items? You will only sample `number_neg_per_pos * sample_size` negative samples.

Reply:
This is to create a dataframe for all the possible interactions between the users and items in the input dataset. Then the existing interactions between users and items in the input data wil be labelled with 1 (positive feedback), whilst the rest will be labelled with 0 (negative feedback). Then for each user, all of the positive feedback and a number of negative feedback (determined by the input `number_neg_per_user`. **NOTE** this is updated, it used to be `number_neg_per_pos` which is not correct).

Reply:
I suspect you could do the sampling in another way so as not to expand the full Cartesian product.
Instead of line 243 below, where you sample from the negative rows of the Cartesian product, you could sample from the items (per user) that do not appear in `df`.
The problem now is that you need O(n^2) memory and CPU. But you could do it in  O(`df.shape[0] + number_neg_per_user * n_users`). We can think about what is the best method available to do this. One way, for example, could be, for each user, to sample `number_neg_per_user` + the number of items in `df`, and then take the setdiff from the items already in `df`.

Reply:
> One way, for example, could be, for each user, to sample number_neg_per_user + the number of items in df, and then take the setdiff from the items already in df.

However, in this case, we don't know which ones are the negative samples, right? And if setdiff does not give us the exact number of negative samples which is pre-determined by `number_neg_per_user`, the operation needs to be done for another time. 

Alternative may be when randomly sample the negative samples, it can be checked against the positive items, and then negative items. This can reduce the complexity to O(df.shape[0] + number_neg_per_user * n_users)

Reply:
For this, can you reuse `user_item_pairs()` in the same module?